src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -299,6 +376,13 @@ private TValue AddOrGetExistingInner(TValue value, out bool addedValue)
             if (value == null)
                 throw new ArgumentNullException();
 
+            if (Interlocked.CompareExchange(ref _entryInProcessOfWritingSentinel, value, null) == null)
+            {
```

Comment:
It may be worth to check if _entryInProcessOfWritingSentinel is already not null before doing CAS.